1. CARGA DE DATOS

1.1. IMPORTACION DE LIBRERIAS

In [228]:
import numpy as np
import pandas as pd

1.2. CARGA DE DATOS

In [229]:
listado = pd.read_csv("C:/Users/sarav/Documents/Cursos/Data analytics The Power/9. Proyecto Final/listingsMADRID.csv")

In [230]:
calendario = pd.read_excel("C:/Users/sarav/Documents/Cursos/Data analytics The Power/9. Proyecto Final/calendarMADRID.xlsx")

2. EXPLORACIÓN INICIAL DE DATOS

2.1. Numero de filas y columas, tipos de datos, primeras filas...

In [231]:
listado.info()

<class 'pandas.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 79 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            25000 non-null  int64  
 1   listing_url                                   25000 non-null  str    
 2   scrape_id                                     25000 non-null  int64  
 3   last_scraped                                  25000 non-null  str    
 4   source                                        25000 non-null  str    
 5   name                                          25000 non-null  str    
 6   description                                   24130 non-null  str    
 7   neighborhood_overview                         10586 non-null  str    
 8   picture_url                                   24999 non-null  str    
 9   host_id                                       25000 non-null  int64  
 1

In [232]:
listado.head()

,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,...,review_scores_communication,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,21853,https://www.airbnb.com/rooms/21853,20250914152907,2025-09-15,previous scrape,Bright and airy room,We have a quiet and sunny room with a good vie...,We live in a leafy neighbourhood with plenty o...,https://a0.muscache.com/pictures/68483181/87bc...,83531,...,4.82,4.21,4.67,NaN,f,2,0,2,0,0.25
1,30320,https://www.airbnb.com/rooms/30320,20250914152907,2025-09-15,city scrape,Apartamentos Dana Sol,NaN,NaN,https://a0.muscache.com/pictures/336868/f67409...,130907,...,4.78,4.90,4.69,NaN,t,17,17,0,0,0.93
2,30959,https://www.airbnb.com/rooms/30959,20250914152907,2025-09-15,previous scrape,Beautiful loft in Madrid Center,Beautiful Loft 60m2 size just in the historica...,NaN,https://a0.muscache.com/pictures/78173471/835e...,132883,...,4.63,4.88,4.25,NaN,f,1,1,0,0,0.06
3,40916,https://www.airbnb.com/rooms/40916,20250914152907,2025-09-15,city scrape,Apartasol Apartamentos Dana,NaN,NaN,https://a0.muscache.com/pictures/hosting/Hosti...,130907,...,4.81,4.88,4.59,NaN,t,17,17,0,0,0.29
4,62423,https://www.airbnb.com/rooms/62423,20250914152907,2025-09-15,city scrape,MAGIC ARTISTIC HOUSE IN THE CENTER OF MADRID,INCREDIBLE HOME OF AN ARTIST SURROUNDED BY PAI...,DISTRICT WITH VERY GOOD VIBES IN THE MIDDLE OF...,https://a0.muscache.com/pictures/miso/Hosting-...,303845,...,4.86,4.97,4.60,NaN,f,3,1,2,0,2.78


In [233]:
listado.shape

(25000, 79)

In [234]:
listado["id"].nunique()

25000

In [235]:
calendario.info()

<class 'pandas.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 7 columns):
 #   Column          Non-Null Count    Dtype         
---  ------          --------------    -----         
 0   listing_id      1048575 non-null  int64         
 1   date            1048575 non-null  datetime64[us]
 2   available       1048575 non-null  str           
 3   price           0 non-null        float64       
 4   adjusted_price  0 non-null        float64       
 5   minimum_nights  1048575 non-null  int64         
 6   maximum_nights  1048575 non-null  int64         
dtypes: datetime64[us](1), float64(2), int64(3), str(1)
memory usage: 56.0 MB


In [236]:
calendario.head()

,listing_id,date,available,price,adjusted_price,minimum_nights,maximum_nights
0,21853,2025-09-15,f,NaN,NaN,4,40
1,21853,2025-09-16,f,NaN,NaN,4,40
2,21853,2025-09-17,f,NaN,NaN,4,40
3,21853,2025-09-18,f,NaN,NaN,4,40
4,21853,2025-09-19,f,NaN,NaN,4,40


El campo comun de ambos es el id y listing_id

In [237]:
calendario.shape

(1048575, 7)

In [238]:
calendario["listing_id"].nunique()

2873

2.2. Valores nulos

In [239]:
listado.isnull()

,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,...,review_scores_communication,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,False,False,False,False,False,False,False,False,False,False,...,False,False,False,True,False,False,False,False,False,False
1,False,False,False,False,False,False,True,True,False,False,...,False,False,False,True,False,False,False,False,False,False
2,False,False,False,False,False,False,False,True,False,False,...,False,False,False,True,False,False,False,False,False,False
3,False,False,False,False,False,False,True,True,False,False,...,False,False,False,True,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,True,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
24995,False,False,False,False,False,False,False,True,False,False,...,True,True,True,False,False,False,False,False,False,True
24996,False,False,False,False,False,False,False,True,False,False,...,True,True,True,False,False,False,False,False,False,True
24997,False,False,False,False,False,False,False,True,False,False,...,True,True,True,False,False,False,False,False,False,True
24998,False,False,False,False,False,False,False,False,False,False,...,True,True,True,False,False,False,False,False,False,True


In [240]:
listado.isnull().sum()

id                                                 0
listing_url                                        0
scrape_id                                          0
last_scraped                                       0
source                                             0
                                                ... 
calculated_host_listings_count                     0
calculated_host_listings_count_entire_homes        0
calculated_host_listings_count_private_rooms       0
calculated_host_listings_count_shared_rooms        0
reviews_per_month                               5147
Length: 79, dtype: int64

Al ser muchas columnas no me salen los nulos de todas, asique lo expreso de otra manera para que me saque las columnas que tienen mas nulos. 

In [241]:
nulos = listado.isnull().sum().sort_values(ascending=False)
nulos.head(20)

calendar_updated               25000
host_neighbourhood             17078
license                        15812
neighbourhood                  14414
neighborhood_overview          14414
host_about                     12402
host_location                   8150
estimated_revenue_l365d         6047
price                           6047
bathrooms                       6040
beds                            6035
review_scores_value             5151
review_scores_location          5150
review_scores_checkin           5148
review_scores_cleanliness       5148
review_scores_communication     5147
reviews_per_month               5147
review_scores_accuracy          5147
first_review                    5147
review_scores_rating            5147
dtype: int64

In [242]:
calendario.isnull()

,listing_id,date,available,price,adjusted_price,minimum_nights,maximum_nights
0,False,False,False,True,True,False,False
1,False,False,False,True,True,False,False
2,False,False,False,True,True,False,False
3,False,False,False,True,True,False,False
4,False,False,False,True,True,False,False
...,...,...,...,...,...,...,...
1048570,False,False,False,True,True,False,False
1048571,False,False,False,True,True,False,False
1048572,False,False,False,True,True,False,False
1048573,False,False,False,True,True,False,False


In [243]:
calendario.isnull().sum()

listing_id              0
date                    0
available               0
price             1048575
adjusted_price    1048575
minimum_nights          0
maximum_nights          0
dtype: int64

2.3. Valores duplicados

In [244]:
listado.duplicated().sum()

np.int64(0)

In [245]:
calendario.duplicated().sum()

np.int64(0)

3. TRANSFORMACION Y LIMPIEZA PROFUNDA

3.1. Identificación de tipos de datos por columnas

In [246]:
listado.dtypes.value_counts()

str        36
float64    23
int64      20
Name: count, dtype: int64

¿Qué columnas string podrian ser realmente numericas o fechas? Mostramos las 35 columnas de texto:

In [247]:
listado.select_dtypes(include="string").columns.tolist()

['listing_url',
 'last_scraped',
 'source',
 'name',
 'description',
 'neighborhood_overview',
 'picture_url',
 'host_url',
 'host_name',
 'host_since',
 'host_location',
 'host_about',
 'host_response_time',
 'host_response_rate',
 'host_acceptance_rate',
 'host_is_superhost',
 'host_thumbnail_url',
 'host_picture_url',
 'host_neighbourhood',
 'host_verifications',
 'host_has_profile_pic',
 'host_identity_verified',
 'neighbourhood',
 'neighbourhood_cleansed',
 'neighbourhood_group_cleansed',
 'property_type',
 'room_type',
 'bathrooms_text',
 'amenities',
 'price',
 'has_availability',
 'calendar_last_scraped',
 'first_review',
 'last_review',
 'license',
 'instant_bookable']

In [248]:
listado[
    [
        "last_scraped",
        "host_since",
        "calendar_last_scraped",
        "first_review",
        "last_review"
    ]
].head()

,last_scraped,host_since,calendar_last_scraped,first_review,last_review
0,2025-09-15,2010-02-21,2025-09-15,2014-10-10,2018-07-15
1,2025-09-15,2010-05-24,2025-09-15,2010-07-06,2025-08-27
2,2025-09-15,2010-05-26,2025-09-15,2015-05-12,2017-05-30
3,2025-09-15,2010-05-24,2025-09-15,2010-11-01,2025-09-11
4,2025-09-15,2010-11-29,2025-09-15,2018-05-10,2025-09-05


Efectivamente hay que cambiarlas a fecha

Comprobación formato columna PRICE

In [249]:
listado["price"].dtype

<StringDtype(storage='python', na_value=nan)>

Comrpobación de porcentajes y booleanos guardados como str

In [250]:
listado[
    [
        "host_response_rate",
        "host_acceptance_rate",
        "host_is_superhost",
        "host_has_profile_pic",
        "host_identity_verified",
        "instant_bookable"
    ]
].head()

,host_response_rate,host_acceptance_rate,host_is_superhost,host_has_profile_pic,host_identity_verified,instant_bookable
0,100%,NaN,f,t,t,f
1,100%,100%,f,t,t,t
2,NaN,NaN,f,t,f,f
3,100%,100%,f,t,t,t
4,100%,99%,f,t,t,f


4.2. Conversión de variables

a)Convertir las columnas str a fecha

In [251]:
columnas_fecha = [
    "last_scraped",
    "host_since",
    "calendar_last_scraped",
    "first_review",
    "last_review"
]

for col in columnas_fecha:
    listado[col] = pd.to_datetime(listado[col])

In [252]:
listado[columnas_fecha].dtypes

last_scraped             datetime64[us]
host_since               datetime64[us]
calendar_last_scraped    datetime64[us]
first_review             datetime64[us]
last_review              datetime64[us]
dtype: object

b) Convertir la columna PRICE a formato numérico

In [253]:
listado["price"] = (
    listado["price"]
    .str.replace("$", "", regex=False)
    .str.replace(",", "", regex=False)
    .astype(float)
)

In [254]:
listado["price"].describe()

count    18953.000000
mean       156.691447
std        492.255350
min          8.000000
25%         70.000000
50%        110.000000
75%        164.000000
max      25654.000000
Name: price, dtype: float64

c)Cambio de str a porcentajes

In [255]:
listado["host_response_rate"] = (
    listado["host_response_rate"]
    .str.replace("%", "", regex=False)
    .astype("float")
)

listado["host_acceptance_rate"] = (
    listado["host_acceptance_rate"]
    .str.replace("%", "", regex=False)
    .astype("float")
)

In [256]:
listado[
    ["host_response_rate", "host_acceptance_rate"]
].describe()

,host_response_rate,host_acceptance_rate
count,20115.000000,20996.000000
mean,90.130748,84.170699
std,22.359971,27.306349
min,0.000000,0.000000
25%,94.000000,84.000000
50%,100.000000,98.000000
75%,100.000000,100.000000
max,100.000000,100.000000


d) Cambio de str a booleanos

In [257]:
columnas_bool = [
    "host_is_superhost",
    "host_has_profile_pic",
    "host_identity_verified",
    "instant_bookable"
]

for col in columnas_bool:
    listado[col] = listado[col].map({
        "t": True,
        "f": False
    })

In [258]:
listado[columnas_bool].dtypes

host_is_superhost         object
host_has_profile_pic      object
host_identity_verified    object
instant_bookable            bool
dtype: object

Al no haberse convertido todo correctamente a bool es porque esas columnas tienen más valores a parte de t y f. Comprobamos:

In [259]:
for col in [
    "host_is_superhost",
    "host_has_profile_pic",
    "host_identity_verified"
]:
    print(f"\n{col}")
    print(listado[col].unique())


host_is_superhost
[False True nan]

host_has_profile_pic
[True False nan]

host_identity_verified
[True False nan]


Salen las opciones True, False y nan. Cambiarlas al booleano que acepte nan

In [260]:
columnas_bool = [
    "host_is_superhost",
    "host_has_profile_pic",
    "host_identity_verified"
]

for col in columnas_bool:
    listado[col] = listado[col].astype("boolean")

In [261]:
listado[columnas_bool].dtypes

host_is_superhost         boolean
host_has_profile_pic      boolean
host_identity_verified    boolean
dtype: object

3.3. Tratamiento de nulos

In [262]:
nulos_pct = (listado.isnull().sum() / len(listado) * 100).sort_values(ascending=False)

nulos_pct.head(15)

calendar_updated             100.000
host_neighbourhood            68.312
license                       63.248
neighbourhood                 57.656
neighborhood_overview         57.656
host_about                    49.608
host_location                 32.600
estimated_revenue_l365d       24.188
price                         24.188
bathrooms                     24.160
beds                          24.140
review_scores_value           20.604
review_scores_location        20.600
review_scores_checkin         20.592
review_scores_cleanliness     20.592
dtype: float64

In [263]:
listado["calendar_updated"].head()

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
Name: calendar_updated, dtype: float64

calendar_updated - 100% nulos - no aporta absolutamente nada, por lo que se puede eliminar

In [264]:
listado["host_neighbourhood"].value_counts().head(10)

host_neighbourhood
Sol              858
Malasaña         627
Embajadores      581
La Latina        520
Cortes           450
Justicia         437
Palacio          330
San Blas         183
Argüelles        175
Ciudad Lineal    158
Name: count, dtype: int64

tiene información importante asique no es tan facil decidir si borrar o no. Tiene info parecida a la columna host_neighbourhood, asique hay que compararlas.

In [265]:
listado["neighbourhood_cleansed"].isnull().sum()

np.int64(0)

como este campo no tiene nulos y dan información parecida, se decide eliminar la columna host_neighbourhood y dejar la de neighbourhood_cleansed

In [266]:
listado["license"].head(10)

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
5    VT-7440
6        NaN
7        NaN
8    VT-6993
9        NaN
Name: license, dtype: str

License es la columna que contiene numeros de licencia turistica. Tiene bastantes nulos (63.2%). Es un dato que no nos va a dar mucha informacion en el analisis, por lo que se decide eliminar.

Me da la impresión de que dos columnas tienen info redundante (neighbourhood y neighbourhood_cleansed). Las comparo:

In [267]:
listado[
    ["neighbourhood", "neighbourhood_cleansed"]
].head(10)

,neighbourhood,neighbourhood_cleansed
0,"Madrid, Spain",Cármenes
1,NaN,Sol
2,NaN,Embajadores
3,NaN,Universidad
4,"Madrid, Spain",Justicia
5,"Madrid, Community of Madrid, Spain",Universidad
6,"Madrid, Spain",Universidad
7,"Madrid, Spain",Universidad
8,NaN,Embajadores
9,NaN,Recoletos


Veo que la primera se refiere mucho a la ciudad/comunidad autonoma y la segunda al barrio.
Voy a ver los valores de la primera:

In [268]:
listado["neighbourhood"].value_counts(dropna=False).head(10)

neighbourhood
NaN                                               14414
Madrid, Comunidad de Madrid, Spain                 9576
Madrid, Community of Madrid, Spain                  688
Madrid, Spain                                       247
Madrid, Co, Spain                                    19
Madrid Center, Madrid, Spain                         10
Madrid , Madrid, Spain                                8
MADRID, Spain                                         7
Chueca, Madrid, Spain                                 5
Pozuelo de Alarcón, Comunidad de Madrid, Spain        2
Name: count, dtype: int64

Decido eliminar la de neighbourhood porque casi todos los valores son de Madrid ciudad, cosa que tiene sentido porque los datos son de ahi. No aporta demasiado.

Revisar otra columna con nulos elevados (57.7%)

In [269]:
listado["neighborhood_overview"].dropna().head(3)

0    We live in a leafy neighbourhood with plenty o...
4    DISTRICT WITH VERY GOOD VIBES IN THE MIDDLE OF...
5    The location of this apartment is ideal for wa...
Name: neighborhood_overview, dtype: str

Son descripciones largas del barrio, por lo que eliminaremos esa columna porque no aporta valor para un analisis. 

Vamos a ver otra columna con nulos elevados (49.6%)

In [270]:
listado["host_about"].dropna().head(3)

0    EN-ES-FR\r\nEN\r\nHi everybody: I'm Abdel. I'm...
1    Apartasol offers a network of several spacious...
2    Estoy empezando en Airbnb y deseo que mis hués...
Name: host_about, dtype: str

Pasa lo mismo que con la anterior asique haremos lo mismo.

Ver si hay columnas con enlaces URL o imagenes:

In [271]:
[
    col for col in listado.columns
    if "url" in col.lower() or "picture" in col.lower()
]

['listing_url',
 'picture_url',
 'host_url',
 'host_thumbnail_url',
 'host_picture_url']

3.4. Eliminar columnas irrelevantes

In [272]:
listado = listado.drop(columns=["calendar_updated"])

In [273]:
listado = listado.drop(columns=["host_neighbourhood"])

In [274]:
listado = listado.drop(columns=["license"])

In [275]:
listado = listado.drop(columns=["neighbourhood"])

In [276]:
listado = listado.drop(columns=["neighborhood_overview"])

In [277]:
listado = listado.drop(columns=["host_about"])

In [278]:
columnas_url = [
    "listing_url",
    "picture_url",
    "host_url",
    "host_thumbnail_url",
    "host_picture_url"
]

listado = listado.drop(columns=columnas_url)

3.5. Tratamiento de nulos de menor porcentaje.
Comprobamos nulos de nuevo:

In [279]:
nulos_pct = (
    listado.isnull().sum()
    / len(listado)
    * 100
).sort_values(ascending=False)

nulos_pct.head(10)

host_location                32.600
price                        24.188
estimated_revenue_l365d      24.188
bathrooms                    24.160
beds                         24.140
review_scores_value          20.604
review_scores_location       20.600
review_scores_cleanliness    20.592
review_scores_checkin        20.592
review_scores_rating         20.588
dtype: float64

In [280]:
listado["host_location"].value_counts(dropna=False).head(15)

host_location
Madrid, Spain                 14582
NaN                            8150
Barcelona, Spain                678
Spain                           153
London, United Kingdom           94
Las Rozas, Spain                 45
Málaga, Spain                    33
Paris, France                    29
Alicante, Spain                  28
Mexico City, Mexico              26
Buenos Aires, Argentina          24
Community of Madrid, Spain       23
New York, NY                     22
Sant Joan d'Alacant, Spain       21
Zarzalejo, Spain                 20
Name: count, dtype: int64

In [281]:
listado["bathrooms"].describe()

count    18960.000000
mean         1.310021
std          0.704779
min          0.000000
25%          1.000000
50%          1.000000
75%          1.500000
max         19.000000
Name: bathrooms, dtype: float64

In [282]:
listado["bathrooms"].head(10)

0    NaN
1    1.0
2    NaN
3    1.0
4    1.5
5    NaN
6    1.0
7    1.0
8    1.0
9    NaN
Name: bathrooms, dtype: float64

In [283]:
[col for col in listado.columns if "bath" in col.lower()]

['bathrooms', 'bathrooms_text']

In [284]:
listado["bathrooms_text"].value_counts().head(15)

bathrooms_text
1 bath              12677
2 baths              3588
1 shared bath        3258
1 private bath       1908
1.5 baths             738
2 shared baths        627
3 baths               511
1.5 shared baths      378
2.5 baths             203
3 shared baths        183
Shared half-bath      154
4 baths               119
0 baths               103
Half-bath              89
0 shared baths         84
Name: count, dtype: int64

Vamos a analizar la columna beds

In [285]:
listado["beds"].describe()

count    18965.000000
mean         1.986291
std          1.523673
min          0.000000
25%          1.000000
50%          2.000000
75%          2.000000
max         40.000000
Name: beds, dtype: float64

In [286]:
listado["beds"].head(10)

0    NaN
1    2.0
2    NaN
3    3.0
4    2.0
5    NaN
6    1.0
7    2.0
8    5.0
9    NaN
Name: beds, dtype: float64

In [287]:
listado[
    ["price", "estimated_revenue_l365d"]
].isnull().sum()

price                      6047
estimated_revenue_l365d    6047
dtype: int64

In [288]:
(
    listado["price"].isnull()
    ==
    listado["estimated_revenue_l365d"].isnull()
).mean()

np.float64(1.0)

In [289]:
listado.loc[
    listado["review_scores_rating"].isnull(),
    "number_of_reviews"
].describe()

count    5147.0
mean        0.0
std         0.0
min         0.0
25%         0.0
50%         0.0
75%         0.0
max         0.0
Name: number_of_reviews, dtype: float64

Se compruba que los valores nulos de las variables de valoracion coinciden con alojamientos que no han recibido aun ninguna reseña (number_of_reviews=0). Por tanto, los nulos de las valoraciones no son un error de calidad del dato. El alojamiento aun no tiene reseñas. Por ello no voy a hacer imputaciones en ninguna de las variables relacionadas con review_scores porque me estaria inventando un dato que no existe

3.6. Tratamiento de nulos del DF CALENDAR

In [290]:
calendario.info()

<class 'pandas.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 7 columns):
 #   Column          Non-Null Count    Dtype         
---  ------          --------------    -----         
 0   listing_id      1048575 non-null  int64         
 1   date            1048575 non-null  datetime64[us]
 2   available       1048575 non-null  str           
 3   price           0 non-null        float64       
 4   adjusted_price  0 non-null        float64       
 5   minimum_nights  1048575 non-null  int64         
 6   maximum_nights  1048575 non-null  int64         
dtypes: datetime64[us](1), float64(2), int64(3), str(1)
memory usage: 56.0 MB


price y adjusted_price tienen un 100% de nulos. En principio podriamos eliminarlos porque no parece que sirvan.

In [291]:
calendario = calendario.drop(columns=["price"])

In [292]:
calendario = calendario.drop(columns=["adjusted_price"])

Vamos a comprobar si la columna available tiene valores booleanos:

In [293]:
calendario["available"].value_counts(dropna=False)

available
f    741670
t    306905
Name: count, dtype: int64

Convertimos a booleano:

In [294]:
calendario["available"] = (
    calendario["available"]
    .map({"t": True, "f": False})
    .astype("boolean")
)

In [295]:
calendario["available"].dtype

BooleanDtype

In [296]:
calendario.isnull().sum()

listing_id        0
date              0
available         0
minimum_nights    0
maximum_nights    0
dtype: int64

3.7. Unión de fuentes - Merge 

In [298]:
dataset = calendario.merge(
    listado,
    left_on="listing_id",
    right_on="id",
    how="left"
)

In [299]:
dataset.shape

(1048575, 73)

Antes de seguir voy a guardar una copia porque este es el dataset principal

In [300]:
dataset.to_csv(
    "airbnb_madrid_merged.csv",
    index=False
)

Auditoria del dataset final - ver nulos

In [301]:
(
    dataset.isnull().sum()
    / len(dataset)
    * 100
).sort_values(ascending=False).head(15)

beds                         44.590516
estimated_revenue_l365d      44.590516
bathrooms                    44.590516
price                        44.590516
host_response_rate           41.214029
host_response_time           41.214029
host_acceptance_rate         37.802732
bedrooms                     21.651289
host_location                14.306559
review_scores_value          13.505949
review_scores_location       13.471139
reviews_per_month            13.436330
review_scores_accuracy       13.436330
review_scores_rating         13.436330
review_scores_cleanliness    13.436330
dtype: float64

review_scores - no lo toco porque son nulos legitimos
host_location - 14% no es excesivo y la variable es interesante como para borrarla
Vamos a ver los demás:

In [303]:
dataset[
    [
        "price",
        "bathrooms",
        "beds"
    ]
].isnull().corr()

,price,bathrooms,beds
price,1.0,1.0,1.0
bathrooms,1.0,1.0,1.0
beds,1.0,1.0,1.0


Esto significa que los nulos aparecen en las mismas filas. Es decir, cuando falta el precio también faltan las camas y los baños.
Si al eliminar esas filas de datos nulos, siguen quedando más de 50.000 filas de datos, creo que lo más adecuado es borrarlas. Borro y compruebo el numero final

In [304]:
dataset_eda = dataset.dropna(
    subset=["price", "bathrooms", "beds"]
)

dataset_eda.shape

(581010, 73)

Con este dataset final doy por terminada la fase de limpieza. He eliminado columnas inutiles, corregido formatos, analizado y justificado los nulos y aun asi queda un archivo grande para analizar.

4. ANALISIS

4.1. Analisis 1: Precio 

In [305]:
dataset_eda["price"].describe()

count    581010.000000
mean        171.225616
std         656.793483
min          15.000000
25%          79.000000
50%         119.000000
75%         176.000000
max       25654.000000
Name: price, dtype: float64

In [306]:
dataset_eda["price"].sort_values(ascending=False).head(20)

109177    25654.0
109178    25654.0
109179    25654.0
109180    25654.0
109149    25654.0
109152    25654.0
109193    25654.0
109194    25654.0
109195    25654.0
109196    25654.0
109165    25654.0
109166    25654.0
109151    25654.0
109136    25654.0
109137    25654.0
109138    25654.0
109139    25654.0
109140    25654.0
109141    25654.0
109142    25654.0
Name: price, dtype: float64

In [307]:
dataset_eda.loc[
    dataset_eda["price"] == dataset_eda["price"].max(),
    "id"
].nunique()

1

Dataset sin duplciados por alojameinto:

In [308]:
alojamientos = dataset_eda.drop_duplicates(subset="id")

In [309]:
alojamientos.shape

(1592, 73)

4.2. Barrios más caros de Madrid:

In [312]:
alojamientos = dataset_eda.drop_duplicates(subset="id")

alojamientos.groupby(
    "neighbourhood_cleansed"
)["price"].agg(["count","mean"]).sort_values(
    "mean",
    ascending=False
)

,count,mean
neighbourhood_cleansed,,
Casa de Campo,5,512.200000
Recoletos,18,359.888889
Sol,153,352.869281
Rios Rosas,10,272.800000
Trafalgar,20,256.750000
...,...,...
Orcasitas,1,40.000000
Canillejas,1,38.000000
San Pascual,2,34.000000


Barrios con minimo 10 alojamientos:

In [313]:
barrios_precio = (
    alojamientos.groupby("neighbourhood_cleansed")["price"]
    .agg(["count","mean"])
)

barrios_precio = barrios_precio[barrios_precio["count"] >= 10]

barrios_precio.sort_values("mean", ascending=False)

,count,mean
neighbourhood_cleansed,,
Recoletos,18,359.888889
Sol,153,352.869281
Rios Rosas,10,272.800000
Trafalgar,20,256.750000
Castellana,16,225.187500
Cortes,83,224.662651
Argüelles,17,203.764706
Justicia,86,196.534884
Comillas,12,173.166667


4.3. Los Superhosts cobran más que los anfitriones normales?

In [314]:
alojamientos.groupby(
    "host_is_superhost"
)["price"].agg(["count","mean"])

,count,mean
host_is_superhost,,
False,930,154.187097
True,613,196.168026


4.4. Habitaciones y precio. ¿Más habitaciones = más precio?

In [315]:
alojamientos.groupby(
    "bedrooms"
)["price"].agg(["count","mean"]).sort_values("mean")

,count,mean
bedrooms,,
1.0,898,108.042316
0.0,127,120.228346
2.0,382,258.693717
3.0,125,265.736000
4.0,40,368.400000
5.0,9,420.777778
7.0,5,678.400000
9.0,1,1000.000000
6.0,5,1065.400000


4.5. Tipo de alojamiento vs precio

In [316]:
alojamientos.groupby(
    "property_type"
)["price"].agg(["count","mean"]).sort_values(
    "mean",
    ascending=False
).head(15)

,count,mean
property_type,,
Entire serviced apartment,62,299.258065
Room in boutique hotel,7,263.714286
Entire home,13,241.923077
Room in hotel,2,218.000000
Entire rental unit,998,201.391784
Room in hostel,13,163.153846
Private room in guest suite,4,154.500000
Entire condo,53,153.698113
Entire hostel,2,153.500000


4.6. Valoraciones vs precio

In [322]:
alojamientos["grupo_valoracion"] = pd.cut(
    alojamientos["review_scores_rating"],
    bins=[0, 4, 4.5, 4.8, 5],
    labels=["Baja", "Media", "Alta", "Excelente"]
)

In [323]:
alojamientos.groupby("grupo_valoracion")["price"].agg(["count","mean"])

,count,mean
grupo_valoracion,,
Baja,23,126.434783
Media,214,136.135514
Alta,747,153.285141
Excelente,582,205.256014


GUARDO ARCHIVO FINAL PARA POWER BI

In [324]:
dataset_eda.to_csv(
    "airbnb_madrid_limpio.csv",
    index=False
)